stepsfunction/handler.py

```diff
@@ -163,19 +178,186 @@ def ocr_page(event, context):
+            workerName=worker_name
+        )
+        LOG.info(f'activityArn={activity_arn}, sf_activity={response}')
+        sf_input = dumps(response['input'])
+        res = sf.send_task_success(
+            taskToken=response["taskToken"],
+            output=sf_input
+        )
         LOG.info(f'sf.send_task_success={res}')
+    except ClientError as e:
+        LOG.error(e)
+
+
+def extract_jpg_from_pdf(bucket, key):
+    """Extract JPG from single-page PDF scan, return as bytes.
+
+    No coversion involved so faster than GhostScript or ImageMagick,
+    and also no loss due to conversion.
+    This mutation of Batchelder's work only handles a single page.
+    Only works with scanned PDF images, not text PDFs.
+    May not always be reliable
+    Past peformance is no guarantee of future results.
+    Use only under a doctor's supervision.
+    """
+    LOG.info(f'Loadding extract_jpg_from_pdf: bucket:{bucket}, key={key}')
+    # download from S3
+    download(bucket, key, '/tmp/page.pdf')
+    pdf = open('/tmp/page.pdf', "rb").read()
+    LOG.info('extract_jpg_from_pdf after open file')
+    i = 0
+    while True:
+        istream = pdf.find(b"stream", i)
+        LOG.info('extract_jpg_from_pdf istream {}'.format(istream))
+        if istream < 0:
+            break
+        istart = pdf.find(STARTMARK, istream, istream + 20)
+        LOG.info('extract_jpg_from_pdf istart {}'.format(istart))
+        if istart < 0:
+            i = istream + 20
+            continue
+        iend = pdf.find(b"endstream", istart)
+        LOG.info('extract_jpg_from_pdf iend {}'.format(iend))
+        if iend < 0:
+            raise Exception("Did not find end of stream!")
+        iend = pdf.find(ENDMARK, iend - 20)
+        if iend < 0:
+            raise Exception("Did not find end of JPG!")
+        istart += STARTFIX
+        iend += ENDFIX
+        LOG.info("JPG from %d to %d" % (istart, iend))
+        jpg = pdf[istart:iend]
+        return jpg
+    raise Exception(f"Could not extract JPG from bucket={bucket}, key={key}")
+
+
+def download(bucket_name, key, path):
+    """Safely download an object from S3.
+
+    param str bucket_name: the name of the S3 bucket
+    param str key: the desired object's S3 key
+    param str path: path on the local disk to save the object to
+    return: None
+    """
+    s3 = get_s3_resource()
+    bucket = s3.Bucket(bucket_name)
+    s3_url = 's3://{}/{}'.format(bucket_name, key)
+    LOG.info('downloading {} to {}'.format(s3_url, path))
+    try:
+        bucket.download_file(key, path)
+    except (ClientError, Exception) as e:
+        LOG.error('download({}, {}, $path) e.response={}'.format(
+            bucket_name, key, e.response))
+        raise e
+
+
+def get_s3_resource():
+    """If the global s3 is set, return it; else set it then return.
+
+    This avoids initializing the s3 connection in every function call so is a
+    performance boost. It's perhaps not as good as doing it in each lambda
```

Comment:
I'd say just get it before the handler and make it a global var. I like calling it "S3C" if it's an S3 client, and S3R if it's an S3 resource.